MoltenVK/MoltenVK/GPUObjects/MVKDevice.mm

```diff
@@ -4786,6 +4789,134 @@ static uint32_t mvkGetEntryProperty(io_registry_entry_t entry, CFStringRef prope
 }
 
 
+#pragma mark Barriers
+
+void MVKDevice::barrierWait(MVKBarrierStage stage, id<MTLCommandBuffer> mtlCommandBuffer, id<MTLRenderCommandEncoder> mtlEncoder, MTLRenderStages beforeStages) {
+	@synchronized (_physicalDevice->getMTLDevice()) {
+		for (auto fence: _activeBarriers[stage]) {
```

Comment:
Vulkan barriers run in submission order, so the fact that this is on MVKDevice (and requires synchronization) worries me
Have you tested what happens if e.g. you encode command buffers in immediate mode and then submit them in the opposite order that you encoded them?  Yes, it won't crash thanks to the `@synchronized` but the fact that this is in a place that requires synchronization at all means that two threads could fight over the _activeBarriers list and probably do unexpected (but non-crashy) things.

Also, any reason you're retaining and releasing all the fences?  Don't they live as long as the MVKDevice (which according to Vulkan should outlive any active work on it)?

Reply:
Right, that's a good point about keeping the fences there, in addition to the multiple queue problem.

Maybe I could avoid requiring to encode only after submit (which would let us keep fences on MVKQueue) by keeping most fences local to the command buffer, and doing some boundary trick to synchronize between submissions on the queue. Not sure what that trick is yet.

The fences are currently only supposed to live as long as the last command buffer that uses them. When one gets removed from all wait/update slots, the only references left are those attached to the command buffer. It sure is more retaining and releasing than I originally expected, so I might just pull the trigger and keep a fixed number of reusable fences..

Reply:
One possibility is to make sure the last group in a submission always updates a known fence, and then always start with waiting on that fence on new submissions:
```
 1   2     3   4     5   6
avb avb B bvc bvc B cva cva
 f   f  B bf  bf  B cf  cf
```
(And if you go the reusable fence route, just have everyone use the same array of fences.  Always start at index 0, and update index 0 at the end of a submission.  Note that fences in Metal, like barriers in Vulkan, also work in submission order, so the worst that could happen using the same fences across multiple encoders at once is more synchronization than you wanted, but assuming you don't mix fences for different pipeline stages, I don't think that will be a big issue.)